### DISCUSSION TITLE
Integrating custom workflows on the websites

### DISCUSSION THREAD

**Discussion by spbreed** [enso-org/enso#7012] [Help and troubleshooting]
Thank you for creating enso. This is truly amazing. 

I'm looking for some help with using Enso to create a custom workflow engine on my website. I want to allow users to select only specific allowed components and allowed to do only specific allowed actions, publish them and run them as scheduled activity.

I've been looking at the Enso documentation, but I'm still not sure how to do this. Can anyone help me out?

Here are some specific questions I have:

* How do I deploy enso workflow engine inside a webapp? 
* How do I restrict users to only selecting specific allowed components?
* How do I restrict users to only doing specific allowed actions?
* How do I publish workflows so that other users can use them?
* How do I run workflows as scheduled activities?

Any help would be greatly appreciated.

**Comment by wdanilo** [comment]
Hi @spbreed, thanks for reaching out! We'd love to help you achieve your goals, but first, I'd need to better understand your use case. Would you be so nice and tell a little bit more of how you want to embed Enso in your web app and what your users should be presented with / allowed to do? What is the high-level idea of this integration? I just want to understand how the thing that you are building will interact with your users. Knowing that, we will be able to provide you with the best hints and guidance :)

**Reply by spbreed** [reply]
Thanks, @wdanilo  The use case is to enable workflow management features for the end users in our web app. Currently, our web app supports inventorying hardware resources, and users can select a resource and perform specific actions such as patching, software updates, running compliance tests, and scheduling, etc., via back-end APIs. Now, the request is to allow customers to create workflows to define a sequence of specific actions and publish them as an template. Once published, these workflows can be scoped to specific resource types, and customers should be able to monitor the state of the workflows.

**Reply by wdanilo** [reply]
Oh, I forgot to write it in my previous response! Regarding `Thank you for creating enso. This is truly amazing.` - thank you so much, it means a lot to us ❤️ 

Thanks for the description, looks like a cool project! It looks like Enso could fit this picture indeed. Although we do not officially support embedding it in other applications yet, this can be done with some work. 

Enso consists of two parts – web-based GUI and a server runtime. The GUI works in all modern browsers (uses WebGL 2.0 to display graphics), while the server runs on JVM. We did not release it yet publicly, but we have here something called "Enso Cloud", which allows you to create workflows online, schedule them, share them with other people, and even create custom components (by selecting a few nodes together and "collapsing them" to a new node) and share them with other users. Would you like to use Enso hosted by us (on AWS) or would you like to host it by yourself? The former option would be much simpler, as in the latter, you'd need to maintain a lot of infrastructure configuration, Enso server spawning, permissions management, data access management, workflow scheduling, etc – all of it is built-into Enso Cloud. It connects to the GUI via web-sockets, so the GUI can be embedded into any web-page (you only need to provide us with a HTML div id and the GUI will spawn there).

Regarding the ability to limit the available components – it's not available out of the box right now, but we can make it work pretty easily. Enso uses libraries of nodes (visual components) under the hood. You can import new libraries and then these nodes will be available in the searcher. By default, we include the standard library containing dozens of high-level and thousands of low-level components, but we can configure it not to display the standard library and display only your own components. It is also possible to theme Enso according to your needs, so you can change its look and feel to match your company needs. We can, of course, help with that :) 

Anyway, I'd love to know more about your company, budget, and timeline in which you want to deliver it to the users. We will be happy to help you create it, highly-tailored for your use case. If that sounds interesting, let's have a chat/call. Are you using Discord? If so, join our server at https://discord.gg/enso and ping me there, I will create a channel for us and include other people to help with this project. After a short chat, it would be good to have a call to discuss it in details. How does it sound to you?

**Reply by spbreed** [reply]
Thank you for the quick updates. This is perfect. Will ping you in discord.

**Reply by wdanilo** [reply]
@spbreed I did not hear back from you - did you write to us via Discord? :)
